databricks/koalas/series.py

```diff
@@ -4772,19 +4772,81 @@ def dot(self, other) -> Union[Scalar, "Series"]:
+        ...     s.dot(kdf)
+        ...
+        x    14
+        y   -14
+        dtype: int64
         """
         if isinstance(other, DataFrame):
-            raise ValueError(
-                "Series.dot() is currently not supported with DataFrame since "
-                "it will cause expansive calculation as many as the number "
-                "of columns of DataFrame"
+            if not same_anchor(self, other):
+                if not self.sort_index().index.equals(other.sort_index().index):
+                    raise ValueError("matrices are not aligned")
+
+                combined = combine_frames(self.to_frame(), other)
+                this_data_spark_column_name = combined["this"]._internal.data_spark_column_names[0]
+                that_data_spark_column_names = combined["that"]._internal.data_spark_column_names
+            else:
+                self_column_label = verify_temp_column_name(other, "__self_column__")
+                combined = DataFrame(
+                    self._internal.with_new_columns(
+                        (
+                            [
+                                self._internal.data_spark_columns[0].alias(
+                                    name_like_string(self_column_label)
+                                )
+                            ]
+                            + other._internal.data_spark_columns
+                        ),
+                        column_labels=([self_column_label] + other._internal.column_labels),
+                    ).resolved_copy
+                )
+                this_data_spark_column_name = name_like_string(self_column_label)
+                that_data_spark_column_names = other._internal.data_spark_column_names
+
+            sdf = combined._internal.spark_frame
+
+            sdf = sdf.select(
+                [
+                    (scol_for(sdf, this_data_spark_column_name) * scol_for(sdf, scol_name)).alias(
+                        scol_name
+                    )
+                    for scol_name in that_data_spark_column_names
+                ]
+            )
+            pdf = (
+                sdf.select(
+                    [
+                        F.sum(scol_for(sdf, scol_name)).alias(scol_name)
+                        for scol_name in that_data_spark_column_names
+                    ]
+                )
+                .toPandas()
+                .transpose()
             )
-        if self._kdf is not other._kdf:
-            if len(self.index) != len(other.index):
-                raise ValueError("matrices are not aligned")
-        if isinstance(other, Series):
+
+            pdf.index = other.columns
+            return cast(Union[Series, Scalar], ks.from_pandas(first_series(pdf)).rename(self.name))
+
+        elif isinstance(other, Series):
+            if not same_anchor(self, other):
+                if not self.sort_index().index.equals(other.sort_index().index):
```

Comment:
@ueshin Shall we match it with pandas' behavior?
In line 4734 of this file, `To match with pandas', it requires to read the whole data for, for example, counting.`
I may not fully understand `requires to read the whole data`, does `sort_index` requires to read the whole data as well?

Reply:
cc @itholic

Reply:
Yeah, actually `self.sort_index().index.equals(other.sort_index().index)` is much more expensive than `len(self.index) != len(other.index)`. We might want to avoid such expensive operations.

Reply:
Would Index.sort_values() be acceptable in this case?

Reply:
I guess it's the same as `self.sort_index().index`.

Reply:
Okay I see.

I restored `series.dot(series)`'s behavior; for `series.dot(dataframe)`, it seems more efforts needed to support non-aligned indexes.

Reply:
To add a bit more to Takuya's comment,

`DataFrame.sort_index()` uses `pyspark.sql.dataframe.DataFrame.sort` internally,

https://github.com/databricks/koalas/blob/138c7b834b5f00d4d96c8ffb272bc1b1e025bd9c/databricks/koalas/frame.py#L6142

and `Index.sort_values()` uses `pyspark.sql.dataframe.DataFrame.orderBy` internally.

https://github.com/databricks/koalas/blob/138c7b834b5f00d4d96c8ffb272bc1b1e025bd9c/databricks/koalas/indexes.py#L1421-L1423

And I just found that these two functions are defined by the same in PySpark.

https://github.com/apache/spark/blob/f94cb53a90558285541090d484a6ae9938fe02e8/python/pyspark/sql/dataframe.py#L1407

Reply:
Great! Thank you so much for looking into this.